### DISCUSSION TITLE
Partitioned but not necessarily ordered results

### DISCUSSION THREAD

**Discussion by soerenwolfers** [duckdb/duckdb#11092] [General]
When switching from duckdb to polars/pandas/numba for numerics-heavy operations, I often need duckdb to give me a dataframe where rows of the same "group" appear next to each other, but I need neither intra- nor inter-group ordering. Is there a way to achieve that?

For example, I may want all rows corresponding to customer A in succession and all rows corresponding to customer B in succession, so I can do O(n) groupwise aggregation or transformation in numba by just recognizing group boundaries; but that doesn't require A coming before B.

In the extreme case of n rows and O(n) groups, achieving what I need via `ORDER BY groupkey` takes O(nlog(n)), while a hash based solution would require O(n) for any number of groups. 

From a quick benchmark I estimate that slows down my code on a dataset with n=1e8 from 1s to 8s.

I think a good syntax for this would be 

```sql
SELECT
  *
FROM
  tbl
ORDER BY 
  groupkey PART
```
because "partitioned-ness" is just "sorted by an order that's not necessarily the default order for the given data type".

**Comment by Tishj** [comment]
I think grouping and ordering are two separate operations already, are you observing different behavior?

**Reply by soerenwolfers** [reply]
I don't mean group as in "GROUP BY". I'm not trying to do aggregation, I just want to rearrange rows such that those in the same "group" are in succession.

**Reply by soerenwolfers** [reply]
@Tishj do you know if this can be achieved? If not, that's fine, I can stop worrying about this then

**Reply by Tishj** [reply]
I have no idea, that's really not an area I'm well versed in

**Comment by soerenwolfers** [comment]
Bump. 

Some motivation by reference to competition (showing what polars can do shows what I'm after and shows it's reasonable to support this in duckdb as well): In polars, this can be achieved via `group_by().all().explode()` (so really, not different from duckdb except the semantics guarantee to preserve ordering in the required way).
Additionally, polars allows "partition by X, then sort within the partitions by Y" via `.group_by(X).all().explode().sort_by(Y).over(X)`
